examples/adversarial-learning/demo_gan_imaging.py

```diff
@@ -0,0 +1,346 @@
+loss_d = adversarial.UnsupAdversarialDiscriminatorLoss(device=device)
+
+
+# %%
+# We are now ready to train the networks using ``AdversarialTrainer``.
+#
+
+trainer = dinv.training.AdversarialTrainer(
+    model=G,
+    D=D,
+    physics=physics,
+    train_dataloader=train_dataloader,
+    eval_dataloader=test_dataloader,
+    epochs=2,
+    losses=loss_g,
+    losses_d=loss_d,
+    optimizer=optimizer,
+    scheduler=scheduler,
+    verbose=True,
+    show_progress_bar=False,
+    save_path=None,
+    device=device,
+)
+G = trainer.train()
+
+
+# %%
+# CSGM / AmbientGAN training
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Compressed Sensing using Generative Models (CSGM) and AmbientGAN are two methods for solving inverse problems
+# using generative models. CSGM (Bora et al., `Compressed Sensing using Generative Models
+# <https://arxiv.org/abs/1703.03208>`__) uses a generative model to solve the inverse problem by optimising the latent
+# space of the generator. AmbientGAN (Bora et al., `AmbientGAN: Generative models from lossy measurements
+# <https://openreview.net/forum?id=Hy7fDog0b>`__) uses a generative model to solve the inverse problem by optimising the
+# measurements themselves. Both methods are trained using an adversarial loss; the main difference is that CSGM requires
+# a ground truth dataset (supervised loss), while AmbientGAN does not (unsupervised loss).
+#
+# In this example, we use a DCGAN as the
+# generator and discriminator, and train using the adversarial loss. The forward pass of the generator is given by:
+#
+# **CSGM** forward pass at train time:
+#
+# .. math:: \hat x = f(z),\quad z\sim \mathcal{N}(\mathbf{0},\mathbf{I}_k)
+#
+# **CSGM**/**AmbientGAN** forward pass at eval time:
+#
+# .. math:: \hat x = f(\hat z)\quad\text{s.t.}\quad\hat z=\operatorname*{argmin}_z \lVert A(f(z))-y\rVert _2^2
+#
+# **CSGM** loss:
+#
+# .. math:: \mathcal{L}=\mathcal{L}_\text{adv}(\hat x, x;D)
+#
+# **AmbientGAN** loss (where :math:`A(\cdot)` is the physics):
+#
+# .. math:: \mathcal{L}=\mathcal{L}_\text{adv}(A(\hat x), y;D)
+#
+# We next load the models and construct losses as defined above.
+
+G = dinv.models.CSGMGenerator(
+    dinv.models.DCGANGenerator(output_size=128, nz=100, ngf=32), inf_tol=1e-2
+)
+D = dinv.models.DCGANDiscriminator(ndf=32)
+_, _, optimizer, scheduler = get_models(
+    model=G, D=D, lr_g=2e-4, lr_d=2e-4
+)  # learning rates from original paper
+
+# For AmbientGAN:
+loss_g = adversarial.UnsupAdversarialGeneratorLoss(device=device)
+loss_d = adversarial.UnsupAdversarialDiscriminatorLoss(device=device)
+
+# For CSGM:
+loss_g = adversarial.SupAdversarialGeneratorLoss(device=device)
+loss_d = adversarial.SupAdversarialDiscriminatorLoss(device=device)
+
+
+# %%
+# As before, we can now train our models. Since inference is very
+# slow for CSGM/AmbientGAN as it requires an optimisation, we only do one
+# evaluation at the end. Note the train PSNR is meaningless as this
+# generative model is trained on random latents.
+#
+
```

Comment:
This would be hard as AmbientGAN is quite tricky to train and I think it would struggle to get good results on this dummy example. I included a pretrained example with DeblurGAN earlier on in the script